### PR TITLE
Add better error handling for permanent redirect responses

### LINKED ISSUE #4458: Using `http://rubygems.org` as the `allowed_push_host` makes `gem push` hang
I followed instructions on https://gist.github.com/jendiamond/6128723 to create a new gem.  When I tried to push it to rubygem.org, I used `bundle exec rake release` and got stuck on:

jz2021_helloworld 0.1.0 built to pkg/jz2021_helloworld-0.1.0.gem.
Tag v0.1.0 has already been created.
Pushing gem to http://rubygems.org...

Since it never says "pushed..." and I never saw the gem on rubygems.org, I assume the push was unsuccessful.  I checked past issues but couldn't find anything matched.  I would really appreciate it if you can provide some insight on what is happening on the Rubygems.org's side since there is no error message returned.

Please let me know if there is anything else I can provide.

Thanks in advance!

### PR DESCRIPTION
**Pull request by jenshenny**
Co-authored-by: Nick Schwaderer <nick.schwaderer@shopify.com>

## What was the end-user or developer problem that led to this PR?
When creating a gem, if someone defines an `allowed_push_host` as `http://rubygems.org` accidentally without `https://` and pushes the gem, they would receive an empty response back. It wasn't until you run the command in verbose until see you see that the response is a permanent redirect and would check the push host defined. @Schwad has experienced this firsthand.

Screen Shot 2022-09-16 at 9 07 57 AM

## What is your fix for the problem, implemented in this PR?

We added a case into `with_response` that returns a better crafted message if the response is a permanent redirect. This new error applies to add/remove/show owners, gem signin and gem pushes.
Screen Shot 2022-09-19 at 9 51 14 AM

For tests I refactored how data for the response is structured since I wanted to mock the location header as well. Instead of an array, I created a Factory that would create Net::HTTPResponse objects.

Closes #4458.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

**Comment by simi** [comment]
@jenshenny ready for review?

**Comment by jenshenny** [comment]
@simi Should be ready now :)

**Review by deivid-rodriguez — APPROVED** [review]
Looks great!

File: test/rubygems/utilities.rb [hunk 1]
```diff
@@ -164,6 +162,30 @@ def download_to_cache(dependency)
...
+#
+#   HTTPResponseFactory.create(
+#     body: "",
+#     code: 301,
+#     msg: "Moved Permanently",
+#     headers: { "location" => "http://example.com" }
+#   )
+#
+
+class HTTPResponseFactory
```
Comment (simi) [inline test/rubygems/utilities.rb:178]: :factory:

**Review by simi — APPROVED** [review]

**Pull request merged**